src/Dispatch.elm

```diff
@@ -0,0 +1,258 @@
+@docs update
+-}
+
+
+import Json.Decode as Json
+import Html.Events
+import Html
+import Dict exposing (Dict)
+import Task
+
+
+{-|
+  Lift any value of type `msg` to a `Cmd msg`.
+-}
+cmd : msg -> Cmd msg
+cmd msg =
+  Task.perform (always msg) (always msg) (Task.succeed msg)
+
+
+
+{-| Message type
+-}
+type Msg m
+  = Forward (List m)
+
+
+type alias Pair m =
+  ( Json.Decoder m, Maybe (Html.Events.Options) )
+
+
+{-| Maps messages to commands
+-}
+forward : Msg m -> Cmd m
+forward (Forward messages) =
+  List.map cmd messages |> Cmd.batch
+
+
+
+{-| Map the second element of a tuple
+
+    map2nd ((+) 1) ("bar", 3) == ("bar", 4)
+-}
+map2nd : (b -> c) -> (a,b) -> (a,c)
+map2nd f (x,y) = (x, f y)
+
+
+{-| Runs batch update
+-}
+update : (a -> b -> ( b, Cmd a )) -> Msg a -> b -> (b, Cmd a)
+update update (Forward msg) model =
+  let
+    inner cmd (m, gs) =
+      let
+        (m', c') = update cmd m
+      in
+        (m', c' :: gs)
+
+  in
+    List.foldl
+      inner
+      (model, [])
+      msg
+    |> map2nd Cmd.batch
+
+
+{-| Applies given decoders to the same initial value
+   and return the applied results as a list
+-}
+applyMultipleDecoders : List (Json.Decoder m) -> Json.Decoder (List m)
+applyMultipleDecoders decoders =
+  let
+    processDecoder initial decoder =
+      case (Json.decodeValue decoder initial) of
+        Ok smt ->
+          Just smt
+
+        Err _ ->
+          Nothing
+  in
+    Json.customDecoder Json.value
+      (\initial ->
+        List.map (processDecoder initial) decoders
+          |> List.filterMap identity
+          |> Result.Ok
+      )
+
+
+{-|
+-}
+forwardDecoder : List (Json.Decoder a) -> Json.Decoder (Msg a)
+forwardDecoder =
+  applyMultipleDecoders >> (Json.map Forward)
+
+
+{-| Run multiple decoders on a single Html Event
+-}
+onEvt :
+  (Msg msg -> msg)
+  -> String
+  -> List (Json.Decoder msg)
+  -> Maybe (Html.Attribute msg)
+onEvt lift event =
+  onEvtOptions lift event Html.Events.defaultOptions
+
+
+{-| Run multiple decoders on a single Html Event with
+the given options
+-}
+onEvtOptions :
+  (Msg msg -> msg)
+  -> String
+  -> Html.Events.Options
+  -> List (Json.Decoder msg)
+  -> Maybe (Html.Attribute msg)
+onEvtOptions lift event options decoders =
+  case decoders of
+    [] ->
+      Nothing
+
+    [ x ] ->
+      Html.Events.onWithOptions event options x
+        |> Just
+
+    _ ->
+      forwardDecoder decoders
+        |> Json.map lift
+        |> Html.Events.onWithOptions event options
+        |> Just
+
+
+onSingle :
+  String
+  -> Html.Events.Options
+  -> List (Json.Decoder msg)
+  -> Maybe (Html.Attribute msg)
+onSingle event options decoders =
+  case decoders of
+    [] ->
+      Nothing
+
+    [ x ] ->
+      Html.Events.onWithOptions event options x
+        |> Just
+
+    -- NOTE: This need to be changed, currently only for debugging
+    x :: xs ->
+      Debug.crash <| "Multiple decoders for Event '" ++ event ++  "' with no `Options.dispatch Mdl`"
```

Comment:
Help me understand how we arrange that this clause is reachable iff `Options.dispatch Mdl` was not used?